middle_end/flambda/types/env/aliases.ml

```diff
@@ -748,7 +749,21 @@ let add_alias t ~element1 ~coercion_from_element2_to_element1 ~element2 =
          as far as [Typing_env] is concerned, but I think this is easier
          to explain than "representative_of_new_alias_class" or some such.) *)
       let alias_of_demoted_element =
-        if Simple.equal element1 canonical_element then element2 else element1
+        (* This needs to return a proper alias of canonical_element, so apply
+           the respective coercion *)
+        if Simple.equal (Simple.without_coercion element1) canonical_element then
+          let coercion_from_element2_to_canonical_element =
+            (* Since canonical_element = canonical_element1 = canonical_element2 *)
```

Comment:
(Minor/ignorable: I would consider moving/duplicating this comment as
it does not apply only to this branch.)